feature: add the auto option for background previews

nealxm opened the pull request:
a start on resolving #46 

for all the other configurations, 0 is used for the auto state. currently this follows that convention for consistency. but i was wondering if 0 might be a state you want to preserve since having 0 background previews may be desirable for lower end hardware.

tildejustin commented:
you could make min = -1 to use for auto, although then you can't use the field (but it seems to be only used for thread priority anyway)

nealxm commented:
-1 made sense to me too yeah. the only issue i found with that is the annotation above the `backgroundPreviews` declaration is for whole numbers only. is there another annotation for integers? i wasn't able to find it

tildejustin commented:
it's just an int, long, and short wrapper, I'm pretty sure it works for negative numbers... https://github.com/KingContaria/SpeedrunAPI/blob/master/src/main/java/org/mcsr/speedrunapi/config/option/WholeNumberOption.java. maybe best to ask contaria.

nealxm commented:
i see, thanks for the help! you're right it works fine subbing -1 for 0. my fault i didn't think the whole number annotation would take integers

Crystal15118 commented:
i was working on it 🙄

Crystal15118 commented:
[This branch](https://github.com/Crystal15118/seedqueue/tree/issue-46) has everything i've done so far

KingContaria commented:
Using rows x columns is not a reliable way of getting the main groups size, since resourcepacks can override that.
Instead `SeedQueueConfig#getBackgroundPreviews` could return an `Optional` and if it's set to AUTO it returns `Optional#empty`. This would then tell the wall screen to calculate the background previews using the data available through `SeedQueueWallScreen#layout`.
That would also allow for including the size of the preparing group in the formula like i suggested in the issue

KingContaria commented:
Also justin is right, -1 will work

nealxm commented:
thanks for your feedback! was working on this a bit with @Crystal15118 and i think this is a step in the right direction. i think it can be done with only `SeedQueueWallScreen#getBackgroundPreviews` without the use of `Optional` & `SeedQueueConfig#getBackgroundPreviews` but i wanted to include that since was suggested

KingContaria commented on src/main/java/me/contaria/seedqueue/gui/wall/SeedQueueWallScreen.java:
```diff
@@ -85,9 +85,10 @@ public class SeedQueueWallScreen extends Screen {
 
     public SeedQueueWallScreen(Screen createWorldScreen) {
         super(LiteralText.EMPTY);
+        this.layout = Layout.createLayout();
```
Creating the layout in the constructor isn't necessary, there is no need to set an initial capacity for the `preparingPreviews` list, just leave it default

Crystal15118 replied:
> Creating the layout in the constructor isn't necessary, there is no need to set an initial capacity for the `preparingPreviews` list, just leave it default

having it empty causes a bug which stops preparing previews from moving into main grid (only the intial main grid seeds are there and after that no more seeds come up there)

KingContaria replied:
That is definitely a different bug, the arraylists initial capacity does not matter for any logic. Per it's Javadoc, "As elements are added to an ArrayList, its capacity grows automatically."

Crystal15118 replied:
> That is definitely a different bug, the arraylists initial capacity does not matter for any logic. Per it's Javadoc, "As elements are added to an ArrayList, its capacity grows automatically."

ok i've tested the second time and seems like you're right

KingContaria commented on src/main/java/me/contaria/seedqueue/gui/wall/SeedQueueWallScreen.java:
```diff
@@ -601,6 +602,15 @@ private List<SeedQueuePreview> getInstances() {
         return instances;
     }
 
+    private int getBackgroundPreviews() {
+        return SeedQueue.config.getBackgroundPreviews().orElseGet(() -> {
+            int mainGroupSize = this.layout.main.size();
+            int preparingGroupSize = Layout.Group.totalSize(this.layout.preparing);
+
+            return Math.min(mainGroupSize + preparingGroupSize, SeedQueue.config.maxCapacity - mainGroupSize);
+        });
+    }
+
```
Please move this method up so it's right below `updatePreparingPreviews`

KingContaria commented on src/main/java/me/contaria/seedqueue/gui/wall/SeedQueueWallScreen.java:
```diff
@@ -403,6 +402,15 @@ private void updatePreparingPreviews() {
+            int mainGroupSize = this.layout.main.size();
+            int preparingGroupSize = Layout.Group.totalSize(this.layout.preparing);
+
+            return Math.min(mainGroupSize + preparingGroupSize, SeedQueue.config.maxCapacity - mainGroupSize);
```
I disagree with `mainGroupSize + preparingGroupSize`, that would always create a larger amount of background previews than the size of the preparing group.
When I mentioned considering the preparing group size, I was thinking the formula I wrote down should be the default, but if the preparing group is bigger than that it should use the size of preparing.
Maybe something like this could work `Math.min(Math.max(mainGroupSize, preparingGroupSize), SeedQueue.config.maxCapacity - mainGroupSize)`.
For example, in a scenario with 12 Max Queued, 4 Main, 6 Preparing it would be `Math.min(Math.max(4, 6), 12) => 6`.
Open to other suggestions tho!

nealxm replied:
i agree with you having `mainGroupSize + preparingGroupSize` isn't the right way to go, having more previews than available positions in the preparing group doesn't make sense. i like the alternative you have provided

nealxm commented:
i added the new process for getting the sensible value and also cleaned up the rest of my edits so it doesn't require the use of `Optional` anymore. hope this is better!

KingContaria commented:
Lgtm, only thing left to discuss is if the option should be renamed to reset all users to AUTO, opinions are welcome

KingContaria commented:
Please rename `backgroundPreviews` (just the field, not the method in `SeedQueueWallScreen`) to reset everyones option to Auto. I'd recommend `preparingPreviews` just because it's consistent with the preparing group in custom layouts.

nealxm commented:
should be good to go!

KingContaria reviewed (approved):
Lgtm!
